core/psh/psh.c

```diff
@@ -1132,17 +1133,20 @@ int main(int argc, char **argv)
 
 	keepidle(1);
 
-	splitname(argv[0], &base, &dir);
+	/* Wait for root filesystem */
+	while (lookup("/", NULL, &oid) < 0)
+		usleep(10000);
 
-	if (!strcmp(base, "psh")) {
-		/* Wait for root filesystem */
-		while (lookup("/", NULL, &oid) < 0)
-			usleep(10000);
+	/* Wait for console */
+	while (write(1, "", 0) < 0)
+		usleep(50000);
 
-		/* Wait for console */
-		while (write(1, "", 0) < 0)
-			usleep(50000);
+	if (psh_login() < 0)
+		return -EACCES;
 
+	splitname(argv[0], &base, &dir);
+
```

Comment:
`splitname()` can change the value of `argv[0]`.
It would probably be better to perform splitname() on copy of `arv[0]`.